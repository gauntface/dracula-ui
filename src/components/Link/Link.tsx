import cx from 'classnames/dedupe'
import { mapValues } from 'lodash'
import React, { HTMLAttributes } from 'react'
import {
  marginMixin,
  MarginMixin,
  paddingMixin,
  PaddingMixin
} from '../../base/spacing'
import { textColors, textSizes, textWeights } from '../Text/Text'

export const linkHoverColors = mapValues(textColors, (classname) => {
  return `${classname}--hover`
})

/** Link Props */
export interface LinkProps
  extends HTMLAttributes<HTMLAnchorElement>,
    PaddingMixin,
    MarginMixin {
  /**
   * Controls the size of the link based on pre-configured Dracula UI sizes.
   * Options: `xsmall`, `small`, `medium`, `large`
   */
  size?: keyof typeof textSizes

  /**
   * Controls the weight of the link.
   * Options: `normal`, `semibold`, `bold`.
   */
  weight?: keyof typeof textWeights

  /**
   * Controls the color of the link
   */
  color?: keyof typeof textColors

  /**
   * Controls the color of the link on hover
   */
  hoverColor?: keyof typeof linkHoverColors
}

/**
 * Text is the base component for any sort of text.
 *
 * Consumers of this component can control, the color, size,
 * weight, and spacing of Text.
 *
 * Use this component for generic, and non-hierarchical text that is
 * to be displayed on a page, or as part of other complex components
 * or UI patterns
 */
export const Link = (props: LinkProps) => {
  const finalProps = {
    ...props,

    className: cx(
      `drac-link`,
      `drac-text`,
      textSizes[props.size ?? 'medium'],
      textWeights[props.weight ?? 'normal'],
      textColors[props.color ?? 'white'],
      linkHoverColors[props.hoverColor ?? 'white'],
      ...paddingMixin(props),
      ...marginMixin(props)
    )
  }

  return React.createElement('a', finalProps, props.children)
}

Link.displayName = 'Link'
