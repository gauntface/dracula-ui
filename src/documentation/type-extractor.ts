import * as ts from 'typescript'

export function extractWithComment(
  fileNames: string[],
  options: ts.CompilerOptions
): void {
  const program = ts.createProgram(fileNames, options)
  const checker: ts.TypeChecker = program.getTypeChecker()

  for (const sourceFile of program.getSourceFiles()) {
    if (!sourceFile.isDeclarationFile) {
      ts.forEachChild(sourceFile, visit)
    }
  }

  function visit(node: ts.Node) {
    const count = node.getChildCount()

    if (count > 0) {
      ts.forEachChild(node, visit)
    }

    if (ts.isInterfaceDeclaration(node) && node.name) {
      const symbol = checker.getSymbolAtLocation(node.name)
      if (symbol) {
        return serializeSymbol(symbol)
      }
    }
  }

  function serializeSymbol(symbol: ts.Symbol) {
    const s = {
      name: symbol.getName(),
      comment: ts.displayPartsToString(symbol.getDocumentationComment(checker)),
      type: checker.typeToString(
        checker.getTypeOfSymbolAtLocation(symbol, symbol.valueDeclaration!)
      )
    }
    console.log(s)
    return s
  }
}

console.log(extractWithComment(['src/components/Avatar/Avatar.tsx'], {}))
