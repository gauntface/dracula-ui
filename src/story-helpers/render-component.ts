import testRenderer from 'react-test-renderer'

export type ComponentExample = {
  code: ReturnType<testRenderer.ReactTestRenderer['toJSON']>
  docs: string
  title: string
}

export type SnapshotBuilder = {
  comp: () => React.ReactElement
  docs: string
  title: string
}

export function buildSnapshot(
  title: string,
  comp: () => React.ReactElement,
  docs: string = ''
): SnapshotBuilder {
  return {
    title,
    comp,
    docs: docs.trim()
  }
}

export function renderSnapshot(
  title: string,
  comp: React.ReactElement,
  docs: string = ''
): ComponentExample {
  const rendered = testRenderer.create(comp)
  const code = rendered.toJSON()
  return {
    title,
    code,
    docs
  }
}
