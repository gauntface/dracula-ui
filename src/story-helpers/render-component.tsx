import React from 'react'
import Enzyme, { mount,  shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import prettier from 'prettier'
import toJson from "enzyme-to-json";

Enzyme.configure({ adapter: new Adapter() })

export type ComponentExample = {
  html: string
  docs: string
  title: string
  react?: any
}

export type SnapshotBuilder = {
  comp: () => React.ReactElement
  docs: string
  title: string
}

export function buildSnapshot(
  title: string,
  comp: () => React.ReactElement,
  docs: string = ''
): SnapshotBuilder {
  return {
    title,
    comp,
    docs: docs.trim()
  }
}

export function renderSnapshot(
  title: string,
  comp: React.ReactElement,
  docs: string = ''
): ComponentExample {

  
  const html = mount(comp).html()
  const react = toJson(shallow(<div>{comp}</div>))

  const prettyHTML = prettier.format(html, {
    parser: 'html'
  })

  return {
    title,
    html: prettyHTML,
    react,
    docs
  }
}
