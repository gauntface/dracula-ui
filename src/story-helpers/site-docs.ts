import {
  renderSnapshot,
  SnapshotBuilder
} from '@/story-helpers/render-component'
import React from 'react'

// import * as serializer from 'jest-snapshot-serializer-raw'

interface Documentation {
  basicUsage: () => SnapshotBuilder
  variations: () => Array<SnapshotBuilder>
}

export function siteDocs<T>(
  componentClass: React.FunctionComponent<T>,
  documentation: Documentation
) {
  describe(`Site: ${componentClass.displayName ?? componentClass.name}`, () => {
    test('Basic Usage', () => {
      const variation = documentation.basicUsage()
      const snapshot = renderSnapshot(
        variation.title,
        variation.comp(),
        variation.docs
      )
      expect(snapshot).toMatchSnapshot()
    })

    const variations = documentation.variations()

    variations.forEach(variation => {
      test(variation.title, () => {
        const snapshot = renderSnapshot(
          variation.title,
          variation.comp(),
          variation.docs
        )
        expect(snapshot).toMatchSnapshot()
      })
    })
  })
}
