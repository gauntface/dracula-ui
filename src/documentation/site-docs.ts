import {
  ComponentExample,
  renderSnapshot,
  SnapshotBuilder
} from '../story-helpers/render-component'
import React from 'react'
import fs from 'fs-extra'
import { pretty } from '../story-helpers/pretty'
import { getDocGen } from '../documentation/docgen/doc-generator'
import { toDSP } from '../documentation/dsp/component-generator'

interface Documentation {
  basicUsage: () => SnapshotBuilder
  variations: () => Array<SnapshotBuilder>
}

export function siteDocs<T>(
  componentClass: React.FunctionComponent<T>,
  documentation: Documentation
) {
  const name = componentClass.displayName ?? componentClass.name

  describe(`Site: ${name}`, () => {
    let examples: Record<string, ComponentExample> = {}

    afterAll(async () => {
      const docGen = getDocGen(name)
      const dsp = toDSP(name, examples, docGen)
      const dspPath = `${process.cwd()}/dsp/data/components/${name}.json`
      await fs.writeFile(dspPath, pretty(JSON.stringify(dsp), 'json'))
    })

    test('Basic Usage', () => {
      const variation = documentation.basicUsage()
      const snapshot = renderSnapshot(
        variation.title,
        variation.comp,
        variation.docs
      )
      expect(snapshot).toMatchSnapshot()
      examples = {
        ...examples,
        basic: snapshot
      }
    })

    const variations = documentation.variations()

    variations.forEach((variation) => {
      test(variation.title, () => {
        const snapshot = renderSnapshot(
          variation.title,
          variation.comp,
          variation.docs
        )
        expect(snapshot).toMatchSnapshot()

        examples = {
          ...examples,
          [variation.title]: snapshot
        }
      })
    })
  })
}
