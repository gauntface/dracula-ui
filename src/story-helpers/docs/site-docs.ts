import {
  ComponentExample,
  renderSnapshot,
  SnapshotBuilder
} from '@/story-helpers/render-component'
import React from 'react'
import fs from 'fs'
import { pretty } from '@/story-helpers/pretty'

interface Documentation {
  basicUsage: () => SnapshotBuilder
  variations: () => Array<SnapshotBuilder>
}

export function siteDocs<T>(
  componentClass: React.FunctionComponent<T>,
  documentation: Documentation
) {
  const name = componentClass.displayName ?? componentClass.name

  describe(`Site: ${name}`, () => {
    let examples: Record<string, ComponentExample> = {}

    afterAll(() => {
      const path = `${process.cwd()}/examples/${name}.json`
      fs.writeFileSync(path, pretty(JSON.stringify(examples), 'json'))
    })

    test('Basic Usage', () => {
      const variation = documentation.basicUsage()
      const snapshot = renderSnapshot(
        variation.title,
        variation.comp,
        variation.docs
      )
      expect(snapshot).toMatchSnapshot()
      examples = {
        ...examples,
        basic: snapshot
      }
    })

    const variations = documentation.variations()

    variations.forEach((variation) => {
      test(variation.title, () => {
        const snapshot = renderSnapshot(
          variation.title,
          variation.comp,
          variation.docs
        )
        expect(snapshot).toMatchSnapshot()

        examples = {
          ...examples,
          [variation.title]: snapshot
        }
      })
    })
  })
}
