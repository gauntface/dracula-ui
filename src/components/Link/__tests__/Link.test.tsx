import { Link } from '@/components/Link/Link'
import { docs } from '@/documentation/site-docs'
import {
  LinkSizings,
  LinkBasic,
  LinkWeights,
  LinkColors
} from '@/components/Link/__stories__/Link.stories'
import { snapshot } from '../../../documentation/render-component'

docs(Link, {
  basic() {
    return snapshot('Usage', LinkBasic)
  },
  variations() {
    return [
      snapshot('Colors', LinkColors),
      snapshot('Weights', LinkWeights),
      snapshot('Sizes', LinkSizings)
    ]
  }
})
