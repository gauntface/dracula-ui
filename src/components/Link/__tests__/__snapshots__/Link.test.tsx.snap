// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Site: Link Basic Usage 1`] = `
Object {
  "docs": "",
  "html": "<a class=\\"drac-link drac-text drac-text-white drac-text-white--hover\\">Link</a>
",
  "react": "<Link>Link</Link>;
",
  "title": "Usage",
}
`;

exports[`Site: Link Colors 1`] = `
Object {
  "docs": "",
  "html": "<div>
  <div style=\\"margin-bottom: 20px\\">
    <a
      color=\\"white\\"
      class=\\"drac-link drac-text drac-text-white drac-text-white--hover\\"
      >white</a
    >
  </div>
  <div style=\\"margin-bottom: 20px\\">
    <a
      color=\\"black\\"
      class=\\"drac-link drac-text drac-text-black drac-text-white--hover\\"
      >black</a
    >
  </div>
  <div style=\\"margin-bottom: 20px\\">
    <a
      color=\\"blackSecondary\\"
      class=\\"drac-link drac-text drac-text-black-secondary drac-text-white--hover\\"
      >blackSecondary</a
    >
  </div>
  <div style=\\"margin-bottom: 20px\\">
    <a
      color=\\"blackLight\\"
      class=\\"drac-link drac-text drac-text-black-light drac-text-white--hover\\"
      >blackLight</a
    >
  </div>
  <div style=\\"margin-bottom: 20px\\">
    <a
      color=\\"cyan\\"
      class=\\"drac-link drac-text drac-text-cyan drac-text-white--hover\\"
      >cyan</a
    >
  </div>
  <div style=\\"margin-bottom: 20px\\">
    <a
      color=\\"green\\"
      class=\\"drac-link drac-text drac-text-green drac-text-white--hover\\"
      >green</a
    >
  </div>
  <div style=\\"margin-bottom: 20px\\">
    <a
      color=\\"orange\\"
      class=\\"drac-link drac-text drac-text-orange drac-text-white--hover\\"
      >orange</a
    >
  </div>
  <div style=\\"margin-bottom: 20px\\">
    <a
      color=\\"pink\\"
      class=\\"drac-link drac-text drac-text-pink drac-text-white--hover\\"
      >pink</a
    >
  </div>
  <div style=\\"margin-bottom: 20px\\">
    <a
      color=\\"purple\\"
      class=\\"drac-link drac-text drac-text-purple drac-text-white--hover\\"
      >purple</a
    >
  </div>
  <div style=\\"margin-bottom: 20px\\">
    <a
      color=\\"red\\"
      class=\\"drac-link drac-text drac-text-red drac-text-white--hover\\"
      >red</a
    >
  </div>
  <div style=\\"margin-bottom: 20px\\">
    <a
      color=\\"yellow\\"
      class=\\"drac-link drac-text drac-text-yellow drac-text-white--hover\\"
      >yellow</a
    >
  </div>
  <div style=\\"margin-bottom: 20px\\">
    <a
      color=\\"purpleCyan\\"
      class=\\"drac-link drac-text drac-text-purple-cyan drac-text-white--hover\\"
      >purpleCyan</a
    >
  </div>
  <div style=\\"margin-bottom: 20px\\">
    <a
      color=\\"yellowPink\\"
      class=\\"drac-link drac-text drac-text-yellow-pink drac-text-white--hover\\"
      >yellowPink</a
    >
  </div>
  <div style=\\"margin-bottom: 20px\\">
    <a
      color=\\"cyanGreen\\"
      class=\\"drac-link drac-text drac-text-cyan-green drac-text-white--hover\\"
      >cyanGreen</a
    >
  </div>
  <div style=\\"margin-bottom: 20px\\">
    <a
      color=\\"pinkPurple\\"
      class=\\"drac-link drac-text drac-text-pink-purple drac-text-white--hover\\"
      >pinkPurple</a
    >
  </div>
  <div style=\\"margin-bottom: 20px\\">
    <a
      color=\\"animated\\"
      class=\\"drac-link drac-text drac-text-animated drac-text-white--hover\\"
      >animated</a
    >
  </div>
</div>
",
  "react": "<div>
  <div style={{ marginBottom: 20 }}>
    <Link color=\\"white\\">white</Link>
  </div>
  <div style={{ marginBottom: 20 }}>
    <Link color=\\"black\\">black</Link>
  </div>
  <div style={{ marginBottom: 20 }}>
    <Link color=\\"blackSecondary\\">blackSecondary</Link>
  </div>
  <div style={{ marginBottom: 20 }}>
    <Link color=\\"blackLight\\">blackLight</Link>
  </div>
  <div style={{ marginBottom: 20 }}>
    <Link color=\\"cyan\\">cyan</Link>
  </div>
  <div style={{ marginBottom: 20 }}>
    <Link color=\\"green\\">green</Link>
  </div>
  <div style={{ marginBottom: 20 }}>
    <Link color=\\"orange\\">orange</Link>
  </div>
  <div style={{ marginBottom: 20 }}>
    <Link color=\\"pink\\">pink</Link>
  </div>
  <div style={{ marginBottom: 20 }}>
    <Link color=\\"purple\\">purple</Link>
  </div>
  <div style={{ marginBottom: 20 }}>
    <Link color=\\"red\\">red</Link>
  </div>
  <div style={{ marginBottom: 20 }}>
    <Link color=\\"yellow\\">yellow</Link>
  </div>
  <div style={{ marginBottom: 20 }}>
    <Link color=\\"purpleCyan\\">purpleCyan</Link>
  </div>
  <div style={{ marginBottom: 20 }}>
    <Link color=\\"yellowPink\\">yellowPink</Link>
  </div>
  <div style={{ marginBottom: 20 }}>
    <Link color=\\"cyanGreen\\">cyanGreen</Link>
  </div>
  <div style={{ marginBottom: 20 }}>
    <Link color=\\"pinkPurple\\">pinkPurple</Link>
  </div>
  <div style={{ marginBottom: 20 }}>
    <Link color=\\"animated\\">animated</Link>
  </div>
</div>;
",
  "title": "Colors",
}
`;

exports[`Site: Link Sizes 1`] = `
Object {
  "docs": "",
  "html": "<div>
  <div style=\\"margin-bottom: 20px\\">
    <a
      class=\\"drac-link drac-text drac-text-lg drac-text-white drac-text-white--hover\\"
      >large</a
    >
  </div>
  <div style=\\"margin-bottom: 20px\\">
    <a class=\\"drac-link drac-text drac-text-white drac-text-white--hover\\"
      >medium</a
    >
  </div>
  <div style=\\"margin-bottom: 20px\\">
    <a
      class=\\"drac-link drac-text drac-text-sm drac-text-white drac-text-white--hover\\"
      >small</a
    >
  </div>
  <div style=\\"margin-bottom: 20px\\">
    <a
      class=\\"drac-link drac-text drac-text-xs drac-text-white drac-text-white--hover\\"
      >xsmall</a
    >
  </div>
</div>
",
  "react": "<div>
  <div style={{ marginBottom: 20 }}>
    <Link size=\\"large\\">large</Link>
  </div>
  <div style={{ marginBottom: 20 }}>
    <Link size=\\"medium\\">medium</Link>
  </div>
  <div style={{ marginBottom: 20 }}>
    <Link size=\\"small\\">small</Link>
  </div>
  <div style={{ marginBottom: 20 }}>
    <Link size=\\"xsmall\\">xsmall</Link>
  </div>
</div>;
",
  "title": "Sizes",
}
`;

exports[`Site: Link Weights 1`] = `
Object {
  "docs": "",
  "html": "<div>
  <div style=\\"margin-bottom: 20px\\">
    <a
      weight=\\"normal\\"
      class=\\"drac-link drac-text drac-text-white drac-text-white--hover\\"
      >normal</a
    >
  </div>
  <div style=\\"margin-bottom: 20px\\">
    <a
      weight=\\"semibold\\"
      class=\\"drac-link drac-text drac-text-semibold drac-text-white drac-text-white--hover\\"
      >semibold</a
    >
  </div>
  <div style=\\"margin-bottom: 20px\\">
    <a
      weight=\\"bold\\"
      class=\\"drac-link drac-text drac-text-bold drac-text-white drac-text-white--hover\\"
      >bold</a
    >
  </div>
</div>
",
  "react": "<div>
  <div style={{ marginBottom: 20 }}>
    <Link weight=\\"normal\\">normal</Link>
  </div>
  <div style={{ marginBottom: 20 }}>
    <Link weight=\\"semibold\\">semibold</Link>
  </div>
  <div style={{ marginBottom: 20 }}>
    <Link weight=\\"bold\\">bold</Link>
  </div>
</div>;
",
  "title": "Weights",
}
`;
